News: GSoC 2022 Work Product BlogPost for Track Suggestion Feature (#282)
mixxxdj/website: content/news/XXXX-XX-XX-gsoc-final-blog-fatih-emre-yildiz.md

##### *Editable Track Menu*
While discussing the `Find On Web` feature on Zulip. There was an idea which was the Track Menu getting crowded and we might consider making it editable. At first I thought the same and this feature would be nice to have and that can help me to get to know the code base for the preference option. So I tried to make it, just to learn how to deal with the preferences. After I introduced this idea on the Zulip channel, that was rejected by the community. So we have decided to not implement this.

Even though we didn't implement this, this taught me that the code base for the preference option (which I was going to use later for the cover art fetcher and the track suggestion feature), I was encountering a Segmentation Fault while implementing this and I have learned to debug with the GNU Debugger, thanks to my mentor. At last, I have learned that `communication is the key` in a community.

##### *Cover Art Fetcher*

> This is the third side project for the Track Suggestion Feature. Fetching cover art from different services. According to the search we have done, we decided to fetch the cover art from CAA (Cover Art Archive). The Cover Art Archive described as on Its web page as "The Cover Art Archive is a joint project between the Internet Archive and MusicBrainz, whose goal is to make cover art images available to everyone on the Internet in an organized and convenient way."

Reviewer: ```suggestion
This is the third side project for the Track Suggestion Feature. Fetching cover art from online resources. According to the search we have done, we decided to fetch the cover art from Cover Art Archive (CAA) hosted by https://archive.org. The Cover Art Archive is a joint project between the Internet Archive and MusicBrainz.
```

> Since the status of the related PR is not merged yet, I will tell you what is on that branch in this section and what is left to do in the Future Work section.

Reviewer: ```suggestion
```
I consider this an unimportant detail. Remove?

> The CAA uses the track's Release MBID (MusicBrainz Identifier) to retrieve cover art. For that, the `Import Metadata From Musicbrainz` could be used. The results returned from the Music Brainz have Release IDs, and by using that we can start to get cover arts from the CAA. To do that, the new network task needed to add, and the design of the `Import Metadata From Musicbrainz` should have been changed.

Reviewer: > the new network task needed to add,

That to much detail for a mews blog.

After long hours of playing, when the floor is crowded the question shows up. "What to play next?". That is a big problem for almost every DJ. This project aims to give suggestions according to tracks playing on the decks. With this feature, Mixxx will have a helper for all the DJs. 

Hey everyone! It's Emre and you've just read my project description. I would like to introduce myself before I dive into the detailed explanation of the blog. I love to listen and play music. I have been always into DJing. That's why I wanted to take a part in Mixxx. I am a person who plays music while chilling-partying with my friends. They always find my music choices are the right fit. But they don't know how difficult to pick the right song, especially after a while of playing. So I thought that would be the best project to work on this summer.

#### Motivation
The track suggestion feature was on the wishlist a long time ago. This feature was mentioned firstly on the [Bug Tracker](https://github.com/mixxxdj/mixxx/issues/6106) and on the [Forum](https://mixxx.discourse.group/t/individual-possible-followers-for-each-track-track-mind/11893). These discussions inspired me to do this project. 

> Before we started to work on this project, I and my mentor planned to do it with the side projects. We choose the side projects so with that, implementation of the big features would have been into pieces and with that we could have a chance to test and review them in smaller environment. These side projects also helped me to get to know the code base, we can think of them just like preparation ground. Also, in one of those side projects we have encountered a bug that took long time to review and test, there was a risk to exceed the time of the project, but luckily time didn't exceed and we could focused on the main part. 

Reviewer 2: ```suggestion
Before we started to work on this project, I and my mentor decided to do it with the side projects. We choose the side projects so with that, implementation of the big features is split into pieces and with that we have a chance to test and review them in smaller environment. These side projects also helped me to get to know the code base, we can think of them just like preparation ground. Also, in one of those side projects we have encountered a bug that took long time to review and test, there was a risk to exceed the time of the project, but luckily we made it and could focus on the main part. 
```

We have come up with a few different side project ideas.

> The first idea was to have a feature called `Find On Web`. This feature aims to help users if they are looking for additional info about a track on online music databases. After this is implemented roughly, we decided to polish it with the `Factory Design Pattern`. Which makes other online music databases easier to implement into codebase. After these two side projects we decided to get cover art of the tracks which is `Cover Art Fetcher`. When I collect tracks for my library, I always wanted to have the correct metadata of the track and also the correct cover art. As a person who cares about appearance a lot, I thought that I am not alone in this situation and people would love it too. That's why I found this side project very good and important! At last, the main project comes along the `Track Suggestion Feature` which is getting Track Suggestions from various services.

Reviewer 2: ```suggestion
The first idea was to have a feature called `Find On Web`. This feature aims to help users if they are looking for additional info about a track on online music databases. After this is implemented roughly, we decided to polish it with the `Factory Design Pattern` which simplifies adding other online music databases later on. After these two side projects we decided to get cover art of the tracks which is `Cover Art Fetcher`. When I collect tracks for my library, I always wanted to have the correct metadata of the track and also the correct cover art. As a person who cares about appearance a lot, I thought that I am not alone in this situation and people would love it too. That's why I found this side project very helpful and important! At last, the main project comes along, the `Track Suggestion Feature` which is getting Track Suggestions from various services.
```

#### Implementation
##### *Find On Web Feature*

> This was the first side project. The `Find On Web`. That helps if the user looking for additional info about a track, album, or artist. 

Reviewer 2: ```suggestion
The first side project `Find On Web` is supposed to help user finding additional info about a track, album, or artist. 
```

> Think about a scenario, you have only just a track from an album and you would like to get more information about other tracks in that album. In normal cases, what you do is, first you Open your favourite browser, then you type the Artist and Album on your search engine, you hit the button to search and you find the web page and finally you can get more information. Think about that you do it very often for each track. That would take your time. This is where the Find On Web comes into hand.

Reviewer 2: ```suggestion
Think about a scenario, you have only just a track from an album and you would like to get more information about other tracks in that album. Until now, you would first open your favourite web browser, then you type the Artist and Album in your search engine, you hit the button to search and you find the web page and finally you can get more information. This is very time-consuming. This is where the Find On Web comes in handy.
```

> Let's think about another scenario. You love to find out remixes of each track in your library and with those remixes you expend your library. You probably use SoundCloud or a similar service very often. Luckily we have this service in our Find On Web feature. If you are looking for Remixes of tracks in your library very often, Find On Web feature will help you along with that.

Reviewer 2: ```suggestion
Let's think about another scenario. You love to find out remixes of each track in your library and with those remixes you expend your library. You probably use SoundCloud or a similar service very often. Luckily we have this service in our Find On Web feature. If you are looking for remixes in your library very often, Find On Web feature will help you along with that.
```

This feature is placed in the Track Menu. When you right-click on a track, you will see different Menus and Actions that you can take over a track. This is called `Track Menu`. On the track menu, in the Metadata section, you can see the Find On Web. There are three main Online Music Databases that exist as menus. They will be called services after this part. These services are:

* SoundCloud
* Discogs
* Last.fm

> When you hover your mouse over those services-menus, you can see the related track properties as actions. Each action populates the suggested Service with the suggested Track Information on a default browser. This looks like:

Reviewer 2: ```suggestion
When you hover your mouse over those services menus, you can see the related track properties as actions. Each action populates the suggested service with the suggested track information for your default browser. This looks like:
```

##### *Find On Web With Factory Design Pattern*
This implementation aims to have a Factory Design Pattern for the Find On Web feature. 

As its name states, you can think about you have a factory and you can produce products. Our factory is the Find On Web itself and our products are the services. The most important point here is that all products coming out of a factory must be derived from the same interface or class. Thanks to this pattern, you can abstract our Services for future use. you can produce products - add new services - straight forward.

Since there are many other online music databases, that is nice to have Find On Web feature with the Factory Design Pattern. Just in case if needed any other service can be added easily without making any changes in the code base.

Related Files:

* [findonwebfactory.cpp](https://github.com/mixxxdj/mixxx/blob/main/src/widget/findonwebmenufactory.cpp)
* [findonwebfactory.h](https://github.com/mixxxdj/mixxx/blob/main/src/widget/findonwebmenufactory.h)
* [Find On Web Menu Services Folder](https://github.com/mixxxdj/mixxx/tree/main/src/widget/findonwebmenuservices)

##### *Editable Track Menu*

> While discussing the `Find On Web` feature on Zulip. There was an idea which was the Track Menu getting crowded and we might consider making it editable. At first I thought the same and this feature would be nice to have and that can help me to get to know the code base for the preference option. So I tried to make it, just to learn how to deal with the preferences. After I introduced this idea on the Zulip channel, that was rejected by the community. So we have decided to not implement this.

Reviewer: You can write something: "The original idea of making the whole Track Menu customizable has been rejected in discussion with the community in favor of good defaults. Sometimes a configuration option just proves the inability of the developers the decide for them. So we have decided to not implement this."

Author: That is way easier to understand and more descriptive at the same time! Thanks.

##### *Cover Art Fetcher*

> This is the third side project for the Track Suggestion Feature. Fetching cover art from online resources. According to the search we have done, we decided to fetch the cover art from Cover Art Archive (CAA) hosted by https://archive.org. The Cover Art Archive is a joint project between the Internet Archive and MusicBrainz.

Reviewer 2: ```suggestion
This is the third side project for the Track Suggestion feature: fetching cover art from online resources. According to the research we have done, we decided to fetch the cover art from Cover Art Archive (CAA) hosted by https://archive.org. The Cover Art Archive is a joint project between the Internet Archive and MusicBrainz.
```

The CAA uses the track's Release MBID (MusicBrainz Identifier) to retrieve cover art. For that, the `Import Metadata From Musicbrainz` feature could be used.

> While Implementing this, At first, I sent the request as described in the API documentation. I was having the response but not correctly, it was saying that the redirected response is received. At first, I tried to solve this issue by sending another request to the redirected URL which was not possible. Then I stopped for a moment and thought about what I can do in this situation, then I found out that QT's class QNetworkAccessManager has [Redirect Policy](https://doc.qt.io/qt-6/qnetworkrequest.html#RedirectPolicy-enum) already. After a long time trying to fix this issue, that was surprising that the solution was in front of me and that was simple.

Reviewer: .. that the solution was in front of me and that was simple.

I think the sentence needs some love. What expect the technical reader what expects a Mixxx user to read?

Author: Okay, I'll take a look.

The new library is an External Library. The located tracks are not available for to play, because they are only text retrieved from Last.fm. But the tracks listed in that library can give an idea to the DJ about what to play next. DJ can see a title or an artist and this can help about picking the next song.

> There is a problem with the Last.fm service, when the track or artist name is misspelled the suggestions can not be fetched. Also for some tracks, there is no suggestion available on the Last.fm servers. To fix this issue, there was another service implementation that was in process, to get the same bpm or key.

Reviewer: How does key and bpm relate to a missing suggestion?

Author: That was related with the getsongbpm, but that shouldn't be on here I guess. But a future reference can be added
